braille-printer-app/brf-printer-app.c

```diff
@@ -0,0 +1,1071 @@
+
+  papplSystemSetVersions(system, (int)(sizeof(versions) / sizeof(versions[0])), versions);
+
+  fprintf(stderr, "brf: statefile='%s'\n", brf_statefile);
+
+  papplSystemSetDNSSDName(system, system_name ? system_name : "brf");
+
+  papplLog(system, PAPPL_LOGLEVEL_INFO, "Auto-adding printers...");
+
+  papplDeviceList(PAPPL_DEVTYPE_USB, (pappl_device_cb_t)printer_cb, system, papplLogDevice, system);
+
+  create_brf_printer(system);
+
+  return (system);
+}
+
+// creating uri for cups-brf printer
+
+int create_brf_printer(pappl_system_t *system)
+{
+  char *dir;
+  struct passwd *pw;
+  int ret;
+
+  // Get the current user's information
+  pw = getpwuid(getuid());
+  if (pw == NULL)
+  {
+    fprintf(stderr, "ERROR: could not get user information\n");
+    return 1;
+  }
+
+  // Create the directory path in the user's home directory
+  if (asprintf(&dir, "%s/BRF", pw->pw_dir) < 0)
+  {
+    fprintf(stderr, "ERROR: could not allocate memory\n");
+    return 1;
+  }
+
+  // Try creating the "BRF" directory with permissions 0700
+  fprintf(stderr, "DEBUG: creating directory \"%s\"\n", dir);
+  ret = mkdir(dir, 0700);
+  if (ret == -1 && errno != EEXIST)
+  {
+    fprintf(stderr, "ERROR: could not create directory \"%s\": %s\n",
+            dir, strerror(errno));
+    free(dir);
+    return 1;
+  }
+  else if (ret == -1)
+  {
+    fprintf(stderr, "DEBUG: directory \"%s\" already exists\n", dir);
+  }
+  else
+  {
+    fprintf(stderr, "DEBUG: directory \"%s\" created successfully\n", dir);
+  }
+
+  // Construct the device URI
+  char device_uri[1024];
+  snprintf(device_uri, sizeof(device_uri), "file://%s", dir);
+  printf("[INFO] Device URI for printer: %s\n", device_uri);
+
+  // Create the printer with the new device URI
+
+  if (papplPrinterCreate(system, 0, "cups-brf", "gen_brf", NULL, device_uri) != NULL)
+  {
+    printf("[SUCCESS] Printer created with device URI: %s\n", device_uri);
+  }
+  else
+  {
+    fprintf(stderr, "[ERROR] Failed to create printer with device URI: %s\n", device_uri);
+  }
+
+  free(dir);
+
+  return 0;
+}
+
+// 'BRFTestFilterCB()' - Print a test page.
+
+// Items to configure the properties of this Printer Application
+// These items do not change while the Printer Application is running
+
+bool // O - `true` on success, `false` on failure
+BRFTestFilterCB(
+    pappl_job_t *job,       // I - Job
+    pappl_device_t *device, // I - Output device
+    void *cbdata)           // I - Callback data (not used)
+{
+  int i;
+  brf_printer_app_global_data_t *global_data = (brf_printer_app_global_data_t *)cbdata;
+
+  brf_cups_device_data_t *device_data = NULL;
+  // brf_job_data_t * job_data;
+  const char *informat;
+  const char *filename;                  // Input filename
+  int fd;                                // Input file descriptor
+  brf_spooling_conversion_t *conversion; // Spooling conversion to use for pre-filtering
+  cups_array_t *spooling_conversions;
```

Comment:
Build with `-Wall -O2`, you will see that the compiler says `spooling_conversions` is not used, you can remove it.

Reply:
As well as others like this